pkg/signup/signup_service.go

```diff
@@ -8,30 +8,75 @@ import (
 	"github.com/codeready-toolchain/registration-service/pkg/kubeclient"
 
 	errors2 "github.com/pkg/errors"
+	"github.com/spf13/viper"
+	apiv1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/validation"
 	"k8s.io/client-go/rest"
 )
 
-// SignupServiceConfiguration represents the config used for the signup service.
-type SignupServiceConfiguration interface {
+// Signup represents Signup resource which is a wrapper of K8s UserSignup
+// and the corresponding MasterUserRecord resources.
+type Signup struct {
+	// The cluster in which the user is provisioned in
+	// If not set then the target cluster will be picked automatically
+	TargetCluster string `json:"targetCluster,omitempty"`
+	// The username.  This may differ from the corresponding Identity Provider username, because of the the
+	// limited character set available for naming (see RFC1123) in K8s. If the username contains characters which are
+	// disqualified from the resource name, the username is transformed into an acceptable resource name instead.
+	// For example, johnsmith@redhat.com -> johnsmith-at-redhat-com
+	Username string `json:"username"`
+	Status   Status `json:"status,omitempty"`
+}
+
+// Status represents UserSignup resource status
+type Status struct {
+	// If true then the corresponding user's account is ready to be used
+	Ready apiv1.ConditionStatus `json:"ready"`
+	// Brief reason for the status last transition.
+	Reason string `json:"reason"`
+	// Human readable message indicating details about last transition.
+	Message string `json:"message,omitempty"`
+}
+
+// ServiceConfiguration represents the config used for the signup service.
+type ServiceConfiguration interface {
 	GetNamespace() string
+	IsTestingMode() bool
+	GetViperInstance() *viper.Viper
 }
 
-// SignupService represents the signup service for controllers.
-type SignupService interface {
+// Service represents the signup service for controllers.
+type Service interface {
+	GetUserSignup(userID string) (*Signup, error)
 	CreateUserSignup(username, userID string) (*crtapi.UserSignup, error)
 }
 
-// SignupServiceImpl represents the implementation of the signup service.
-type SignupServiceImpl struct {
-	Namespace   string
-	UserSignups kubeclient.UserSignupInterface
+// ServiceImpl represents the implementation of the signup service.
+type ServiceImpl struct {
+	Namespace         string
+	UserSignups       kubeclient.UserSignupInterface
+	MasterUserRecords kubeclient.MasterUserRecordInterface
+	checkerFunc       func(userID string) (*Signup, error)
 }
 
-// NewSignupService creates a service object for performing user signup-related activities
-func NewSignupService(cfg SignupServiceConfiguration) (SignupService, error) {
+// NewSignupService creates a service object for performing user signup-related activities.
+func NewSignupService(cfg ServiceConfiguration) (Service, error) {
+
+	if cfg.IsTestingMode() {
```

Comment:
You don't need any checkerFunc or test code here at all. I committed a small change which replace the client by the MUR interface: https://github.com/codeready-toolchain/registration-service/pull/41/commits/546d71df03959d41c9973d984ee4164afe10074a

All you need in your test is:
```
svc, err := NewSignupService(cfg)
svc.UserSignups = mockUserSignupsImpl
svc.MasterUserRecords = mockMURImpl
```

Reply:
I feel that that this is all a bit overengineered. What I want to do here is to seperate the test and logic used for the controller from the logic in the service (or anything in the kube layer). This will seperate the tests as well. Currently, the test for the controller focusses solely on the controller code test, mocking the whole kube layer. 

You are suggesting to "install" the mock a layer further down in the kube layer, but that makes the testing of the controller code much more complex as it has to deal with the details of how the result from the kube api comes together. 

I don't think this is good. Yes, it will remove a few "extra turns" done for the test (like the checkerFunc mechanism), but it diverts the controller test and makes this very complex. Plus, the timed test is no longer possible without a lot of extra work.

Reply:
And this also makes the middleware test much more complex as this one now also needs to mock that complete k8s client shebang. The signupService is added in the routes setup, so it is part of all tests that run through the routes setup (like the middleware test).